File: Data Structures & Algorithms/validate-parentheses/submission-0.js
class Solution {
    /**
     * @param {string} s
     * @return {boolean}
     */
    isValid(s) {

        let stack = [];
        
        for(let ch of s){
            let top = stack[stack.length - 1];
            if(ch === '(' || ch === '{' || ch === '[') stack.push(ch);
            else if((ch === ')' || ch === '}' || ch === ']') && stack.length === 0) return false;
            else if((ch === ')' && top !== '(') || (ch === '}' && top !== '{') || (ch === ']' && top !== '[')) return false;
            else stack.pop();
        }

        if(stack.length === 0) return true;
        else return false;
    }
}

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<pair<int, int>> st; // stores {height, index}
        int maxArea = 0;

        for (int i = 0; i < heights.size(); i++) {
            int index = i;
            while (!st.empty() && st.top().first > heights[i]) {
                index = st.top().second;
                maxArea = max(maxArea, st.top().first * (i - st.top().second));
                st.pop();
            }
            st.push({heights[i], index});
        }

        while (!st.empty()) {
            maxArea = max(maxArea, st.top().first * (int(heights.size()) - st.top().second));
            st.pop();
        }

        return maxArea;
    }
};
